julearn/transformers/confounds.py

```diff
@@ -14,21 +14,31 @@ def __init__(
         self, model_confound=None, confounds='use_suffix',
         threshold=None, suffix='__:type:__confound'
     ):
-        """Model to regress out/remove confound
-        by a given model from the features.
+        """Transformer which can use pd.DataFrames and remove the confounds
+        from the features by subtracting the predicted features
+        given the confounds from the actual features.
 
         Parameters
         ----------
-        model_confound : [sklearn model]
-                         Model by which we predict X
-                         given the confound to later compute the residuals.
-        confounds   : [int, str]
-                         The indx in ndarray or
-                         the column name in a df of the confound.
-
-        threshold      : [float]
-                         Threshold under which all values are rounded to 0.
-
+        model_confound : sklearn.base.BaseEstimator
+            Model used to predict all features independentley
+            using the confounds as features. The predictions of these models
+            are then subtracted from each feature,
+            default LinearRegression()
+        confounds : list[str] or str
```

Comment:
While I like the idea (also mentioned in #19). I do prefer removing the `suffix` and `use_suffix` approach for a more automatic one.

The idea is that instead of using the strings as literals, pass them through a regex match first. So it's not 
```
to_use = df[confounds]
```

but
```
columns = []
for c in confounds:
    re.compile(x)
    columns.extend([re.search(x) for x in df.columns]
to_use = df[columns]
```
So confounds can be a string or list of strings that are applied using a regex. Same for X.

To implement the default behaviour implemented here: `confounds = '*__:type:__confound'`

I don't say it should be changed here, but when #19 is implemented like this, this class will change. So it's better to keep the public API as consistent as possible between versions.

Reply:
I am thinking about it, but just to give you the complete context.
For the user api you can make this easily possible without changing this class or the pipeline.

But I am concerned about the back-end part.
There should be a consistency throughout the back-end. 
Currently, we use a `column_type_sep` e.g. `__:type:__` to  split the column_names from the column_types (e.g. confound, categorical etc.). 
I find this very predictable, which makes edge cases less prominent I think. 
If you want to use regex instead of `suffix`, which simplifies the just `column_type_sep` then we have to do it across the complete back-end. So, also in the `pipeline`. 
Here, I am a bit afraid that this might lead to more problematic edge cases.
But its no problem and I am exaggerating.  

So, I could only change the confound removal to use regex. 
It would still work with the pipeline using the default regex you mentioned. 
But it might get a bit inconsistent with the idea how the pipeline determines the column_type
I can change everything to regex, but I am not sure whether this might lead to some problems with strange user input. 
Or we just keep it and change it in the user api. 
Honestly, for pp.l using the confound removal class on its own and want to use regex:
They could use the regex to get the columns themselves and give it as a list to the confound removal class.

Reply:
I'm not addressing the `__:type:__` part. This is fine. I agree that it's a good idea to keep it encoded with the datafrma.e 

But I see that there's a complexity issue if you want to use something more than the standard pipeline/confound removal. Let's say you want to understand what the `ExtendedDataFramePipeline` and `DataFrameConfoundRemover` do in order to set up a few parameters.

This:
```Python
    def __init__(
        self, model_confound=None, confounds='use_suffix',
        threshold=None, suffix='__:type:__confound'
    ):
```

Could be redefined into this:
```Python
    def __init__(
        self, model_confound=None,
        confounds='*__:type:__confound',
        threshold=None
    ):
```

And then, in the pipeline, remove `column_type_sep`:
```Python
    def __init__(self, dataframe_pipeline,
                 y_transformer=None,
                 confound_dataframe_pipeline=None,
                 confounds=None, categorical_features=None,
                 return_trans_column_type=False):
```

The main argument is while having these parameters provide flexibility, we are leaving to the user to match stuff across classes. Having a constant `__:type:__` across the backend does not allow for the user to do wrong things.

If you are also not into using regex, you can even remove `confounds` as a parameter for `DataFrameConfoundRemover`:
```Python
    def __init__(self, model_confound=None, threshold=None):
```

And as requirement, it will consider as confounds all the columns whose name ends like `__:type:__confound`. But my impression is that this might be too restrictive.

The best way to figure this out is to write advanced examples and check if there is something you can't. That's why I am insisting with the examples.

If you can't think of an example where one parameter is used, then this parameter should not be there.

Reply:
The idea behind making a suffix possible was in case people already use one in there data.
But I misunderstood your points. You want to remove the option of changing the the separator from __:type:__ to smth else right ?  I can do this reducing the complexity is a good idea. 

About the regex. From how I understand you you only want to have the option of regex in the cofound_removal step. 
In case ppl. want to use it outside of the api. 

Then I would still propose 3 options: 
0.`*__:type:__confound`  default for when data follows `__:type:__` convention
1. another regex 
2. a list of explicit column names. <- I would not get rid of this option. 
I think it is the easiest to use it outside of our api system

 If I understood you correctly, I would just change this and the test accordingly.

Reply:
My take is that if people use already a suffix in their data and want to _hack_ into the internals of julearn, they can adapt their data. I would not worry against _hackers_.

The beauty of using regex is that a literal also matches:
```Python
columns = ['a_feat_a', 'b_feat_b', 'c_feat_c', 'd_conf_d', 'e_conf_e']
confounds = ['d_conf_d', 'e_conf_e']
```
This will be the code to convert `confounds` to the column names to use (`picks`):
```Python
picks = []
for col in columns:
    if any(re.search(exp, col) for exp in confounds):
        picks.append(col)
print(picks)
```

Will have `picks == confounds`.
It can be also specified like this:
```Python
columns = ['a_feat_a', 'b_feat_b', 'c_feat_c', 'd_conf_d', 'e_conf_e']
confounds = ['.*_conf_.*']
```

So in julearn.utils, we can define
```Python
def pick_columns(regexes, columns):
    if not isisntance(regexes, list):
        regexes = [regexes]
    picks = []
    for col in columns:
        if any(re.search(exp, col) for exp in regexes):
            picks.append(col)
```

And at the beginning of everything that accept X or confounds as string or list of strings, we use that function to find the actual column names. This way, we can accept the literal names of the columns or a list of regular expressions.

By the way, the default should be `.*__:type:__confound` (mind the . before the *)

Reply:
If you are happy with the code and CI is ok, LGTM!